File: tools/bench/answers/go_sonnet/s3/14_rect_lib.go
package main

import "fmt"

type rect struct {
	width  float64
	height float64
}

func area(r rect) float64 {
	return r.width * r.height
}

func perimeter(r rect) float64 {
	return 2 * (r.width + r.height)
}

func scale(r rect, factor float64) rect {
	return rect{r.width * factor, r.height * factor}
}

func main() {
	r := rect{3.0, 4.0}
	fmt.Println(area(r))
	fmt.Println(perimeter(r))
	fmt.Println(area(scale(r, 2.0)))
}

File: tools/bench/answers/rust_sonnet/s3/14_rect_lib.rs
struct Rect {
    width: f64,
    height: f64,
}

fn area(r: &Rect) -> f64 {
    r.width * r.height
}

fn perimeter(r: &Rect) -> f64 {
    2.0 * (r.width + r.height)
}

fn scale(r: &Rect, factor: f64) -> Rect {
    Rect { width: r.width * factor, height: r.height * factor }
}

fn main() {
    let r = Rect { width: 3.0, height: 4.0 };
    println!("{}", area(&r));
    println!("{}", perimeter(&r));
    let r2 = scale(&r, 2.0);
    println!("{}", area(&r2));
}
